File: lib/item.rb
# require_relative './base_entity'

class Item < Errol::Entity
  require_relative './item/record'
  def self.repository
    Items
  end
  
  entry_accessor  :name,
                  :initial_price,
                  :discounted_price,
                  :asset

  def piece
    Piece.new record.piece_record if record.piece_record
  end

  def piece=(piece)
    record.piece_record = piece.record
  end

  def subsequent_price
    discounted_price || initial_price
  end

  def multibuy_discount?
    !!discounted_price
  end

  def price_for(n)
    initial_price + subsequent_price * (n-1)
  end

end

File: apps/catalogue/lib/catalogue/item.ex
defmodule UM.Catalogue.Item do
  defstruct [
    id: nil,
    name: nil,
    initial_price: nil,
    discounted_price: nil,
    asset: nil,
    piece_id: nil
  ]

  def price_for(item, n) when is_integer(n) and n > 0 do
    Money.add(item.initial_price, Money.multiply(subsequent_price(item), (n-1)))
  end

  def initial_price(%{initial_price: i}) do
    i
  end

  def subsequent_price(%{initial_price: i, discounted_price: d}) do
    d || i
  end

  def multibuy_discount?(item) do
    !!item.discounted_price
  end
end
